### ISSUE TITLE
Unicode errors on string columns with NaNs

### ISSUE THREAD

**Issue by epervago** [dask/dask#1883]
Using `dd.from_pandas` on a Pandas dataframe with a unicode or UTF-8 text column with some missing values (NaNs) results in UnicodeEncodeError. Minimal test case:

```python
import pandas as pd
import numpy as np
import dask.dataframe as dd

df = pd.DataFrame({'x': [u'foo', u'Jos\xe9', np.nan]})
dd.from_pandas(df, npartitions=2)
```
Looking at `normalize_array` in https://github.com/dask/dask/blob/7a2d5e378fe67b54df74b0b52ca160a7f3952513/dask/base.py#L330 the first case fails since join doesn't work on nans and the second one fails because it tries to convert it to str (bytes in Python 3), resulting in the unicode error.

**Comment by mrocklin** [comment]
Verified that this fails in Python 2.  It does appear to operate fine in Python 3.  Thanks for reporting and the clear and concise test case.

**Comment by epervago** [comment]
You're very welcome. Here's a commit on my fork that fixes it (tested on Python 2 and 3):

https://github.com/epervago/dask/commit/958732ce6c51ef6af39db4727d948bfa66a0a8d6

**Comment by mrocklin** [comment]
Cool.  Are you interested in submitting a pull request?  If so it would be good to also add a test (your example above would work well) and to verify that this doesn't significantly impact performance.

**Comment by JohnCrickett** [comment]
Whist looking at this I saw that test_encoding_gh601 in dask/dataframe/io/tests/test_csv.py failed with all three parameters under Python 2.7.13. 

Python 3.5.2 was fine.

**Comment by jcrist** [comment]
Fixed by #1945. Closing.
